ssl/t1_enc.c

```diff
@@ -316,6 +325,63 @@ int tls1_change_cipher_state(SSL *s, int which)
                  ERR_R_INTERNAL_ERROR);
         goto err;
     }
+#ifndef OPENSSL_NO_KTLS
+    if (s->compress)
+        goto skip_ktls;
+
+    if ((which & SSL3_CC_READ) ||
+        ((which & SSL3_CC_WRITE) && (s->mode & SSL_MODE_NO_KTLS_TX)))
+        goto skip_ktls;
+
+    /* ktls supports only the maximum fragment size */
+    if (ssl_get_max_send_fragment(s) != SSL3_RT_MAX_PLAIN_LENGTH)
+        goto skip_ktls;
+
+    /* check that cipher is AES_GCM_128 */
+    if (EVP_CIPHER_nid(c) != NID_aes_128_gcm
+        || EVP_CIPHER_mode(c) != EVP_CIPH_GCM_MODE
+        || EVP_CIPHER_key_length(c) != TLS_CIPHER_AES_GCM_128_KEY_SIZE)
+        goto skip_ktls;
+
+    /* check version is 1.2 */
+    if (s->version != TLS1_2_VERSION)
+        goto skip_ktls;
+
+    wbio = s->wbio;
+    if (!ossl_assert(wbio != NULL)) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, SSL_F_TLS1_CHANGE_CIPHER_STATE,
+                 ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+
+    /* All future data will get encrypted by ktls. Flush the BIO or skip ktls */
+    if (BIO_flush(wbio) <= 0)
+        goto skip_ktls;
+
+    /* ktls doesn't support renegotiation */
+    if (BIO_get_ktls(s->wbio))
+        goto err;
```

Comment:
I tried a renegotiation from s_client to see what would happen. I got an assertion failed error and abort! This error needs to call `SSLfatal` first. We later do a sanity check to make sure we've sent an alert via SSLfatal if we're in an error state. That assertion was failing in this case (also see #7594). I wonder if we should automatically set `SSL_OP_NO_RENEGOTIATION` if ktls is in use.

Reply:
...or possibly explicitly check if we're using ktls everywhere we check `SSL_OP_NO_RENEGOTIATION`

Reply:
Fixed the compilation issue.
Added the call to SSLFatal, and set the SSL_OP_NO_RENEGOTATION option on the SSL object.